opts.go

```diff
@@ -55,44 +80,149 @@ func WithLoNetwork() ConfigOptions {
+		if err != nil {
+			return err
+		}
+		confList, err := cnilibrary.ConfListFromConf(conf)
+		if err != nil {
+			return err
+		}
+		c.Lock()
+		defer c.Unlock()
+		c.networks = append(c.networks, &Network{
+			cni:    c.cniConfig,
+			config: confList,
+			ifName: getIfName(c.prefix, 0),
+		})
 		return nil
 	}
 }
 
-//TODO: Should we support direct network configs?
-/*
-func WithConf(byte []bytes) ConfigOptions {
-	return func(c *config) error {
-			c.networks=
-	}
-}
-
-func WithConfFile(fileName string) ConfigOptions {
-	return func(c *config) error {
-
+// WithConfFile can be used to load network config
+// from an .conf file. Supported with absolute fileName
+// with path only.
+func WithConfFile(fileName string) LoadOptions {
+	return func(c *libcni) error {
+		conf, err := cnilibrary.ConfFromFile(fileName)
+		if err != nil {
+			return err
+		}
+		// upconvert to conf list
+		confList, err := cnilibrary.ConfListFromConf(conf)
+		if err != nil {
+			return err
+		}
+		c.Lock()
+		defer c.Unlock()
+		c.networks = append(c.networks, &Network{
+			cni:    c.cniConfig,
+			config: confList,
+			ifName: getIfName(c.prefix, 0),
+		})
+		return nil
 	}
 }
 
-func WithConfList(byte []bytes) ConfigOptions {
-	return func(c *config) error {
-      c.
+// WithConfListFile can be used to load network config
+// from an .conflist file. Supported with absolute fileName
+// with path only.
+func WithConfListFile(fileName string) LoadOptions {
+	return func(c *libcni) error {
+		confList, err := cnilibrary.ConfListFromFile(fileName)
+		if err != nil {
+			return err
+		}
+		c.Lock()
+		defer c.Unlock()
+		c.networks = []*Network{&Network{
+			cni:    c.cniConfig,
+			config: confList,
+			ifName: getIfName(c.prefix, 0),
+		},
+		}
+		return nil
 	}
 }
-func WithConfListFile(files []string) ConfigOptions {
-	return func(c *config) error {
 
+// WithDefaultConf can be used to detect network config
+// files from the configured cni config directory and load
+// them.
+func WithDefaultConf() LoadOptions {
+	return func(c *libcni) error {
+		files, err := cnilibrary.ConfFiles(c.pluginConfDir, []string{".conf", ".conflist", ".json"})
+		switch {
+		case err != nil:
+			return errors.Wrapf(ErrRead, "failed to read config file: %v", err)
+		case len(files) == 0:
+			return errors.Wrapf(ErrCNINotInitialized, "no network config found in %s", c.pluginConfDir)
+		}
+
+		// files contains the network config files associated with cni network.
+		// Use lexicographical way as a defined order for network config files.
+		sort.Strings(files)
+		// Since the CNI spec does not specify a way to detect default networks,
+		// the convention chosen is - the first network configuration in the sorted
+		// list of network conf files as the default network and choose the default
+		// interface provided during init as the network interface for this default
+		// network. For every other network use a generated interface id.
+		i := 0
+		c.Lock()
+		defer c.Unlock()
```

Comment:
Not sure what the concurrence story is for this, but this could block for a long time, what about only locking when needed inside the loop?

Reply:
The concurrency story is we could have parallel calls to Status , Load (read kubernetes), its just safer to protect the instance of networks. I don't expect this to take long. Typically config file is one in kube and its a one time op.